### PR TITLE
Refactors WrapperQueryBuilder and Parser

### PR DESCRIPTION
**Pull request by alexksikes**
Relates to #10217

This PR is against the query-refactoring branch.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java [hunk 1]
```diff
@@ -67,23 +73,94 @@ public WrapperQueryBuilder(byte[] source, int offset, int length) {
...
+    public WrapperQueryBuilder(byte[] source) {
+        this(source, 0, source.length);
+    }
+
+    /**
      * Creates a query builder given a query provided as a {@link BytesReference}
      */
     public WrapperQueryBuilder(BytesReference source) {
-        this.source = source.array();
+        this.source = source.toBytes();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: why was this changed?

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java [hunk 2]
```diff
@@ -67,23 +73,94 @@ public WrapperQueryBuilder(byte[] source, int offset, int length) {
...
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
         builder.field("query", source, offset, length);
+        printBoostAndQueryName(builder);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: this was not there on purpose, it doesn't really make sense for this wrapper to have its own boost and query _name I think. I was considering adding no-op setters for those like I did in `QueryFilterBuilder`

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java [hunk 3]
```diff
@@ -67,23 +73,94 @@ public WrapperQueryBuilder(byte[] source, int offset, int length) {
...
+    @Override
+    protected WrapperQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        return new WrapperQueryBuilder(in.readByteArray(), in.readVInt(), in.readVInt());
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeByteArray(this.source);
+        out.writeVInt(this.offset);
+        out.writeVInt(this.length);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: can't we just serialize the byte array only, only the part that we are interested in and skip the rest?

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java [hunk 4]
```diff
@@ -41,28 +39,37 @@ public WrapperQueryParser() {
...
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if ("query".equals(currentFieldName)) {
+                    source = parser.binaryValue();
+                } else if ("boost".equals(currentFieldName)) {
+                    boost = parser.floatValue();
+                } else if ("_name".equals(currentFieldName)) {
+                    queryName = parser.text();
+                } else {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java]: I think we should go back to not parsing boost and queryname here, it's an exception but in the context of this wrapper, I think only the inner ones make sense, not boost and _name set on the wrapper itself

File: core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    @Override
+    protected WrapperQueryBuilder doCreateTestQueryBuilder() {
+        QueryBuilder wrappedQuery = RandomQueryBuilder.createQuery(random());
+        switch (randomInt(2)) {
+            case 0:
+                return new WrapperQueryBuilder(wrappedQuery.toString());
+            case 1:
+                return new WrapperQueryBuilder(wrappedQuery.buildAsBytes().toBytes());
+            default:
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java]: make it case 2?

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java [hunk 6]
```diff
@@ -67,23 +73,94 @@ public WrapperQueryBuilder(byte[] source, int offset, int length) {
...
+        return this.source;
+    }
+
+    public int offset() {
+        return this.offset;
+    }
+
+    public int length() {
+        return this.length;
+    }
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: I wonder if it's worth/possible to skip these getters in this case

**Comment by javanna** [comment]
left some comments

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java (hunk 3 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: we could but then we'd have to forget about this public constructor WrapperQueryBuilder(byte[] source, int offset, int length) or make source a BytesArray.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java (hunk 6 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: only offset() and length() can be removed.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java (hunk 4 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java]: sounds good, thanks.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java (hunk 2 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: thanks, will do.

**Comment by alexksikes** [comment]
@javanna Thanks for the review. I updated the PR.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java (hunk 3 above)
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: Why not keep the constructor as is, but on serialisation only write out the part of the byte array that we are interested in (and on reading it initalise offset to zero and length to the length of the byte array we've read) like we already do in doXContent?

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java [hunk 7]
```diff
@@ -71,19 +70,74 @@ public WrapperQueryBuilder(byte[] source, int offset, int length) {
...
+    @Override
+    public String getName() {
+        return NAME;
     }
 
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
-        builder.field("query", source, offset, length);
+        builder.field("query", source);
```
Comment (MaineC) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: See the comment I added to the original discussion on the outdated diff.

**Comment by MaineC** [comment]
Added some comments.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java (hunk 3 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: Yes I actually removed this constructor, because this is what is happening anyway under the hood. This means that the vars offset and length have been removed.

**Comment by alexksikes** [comment]
@MaineC Thanks I have addressed your comment.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java [hunk 8]
```diff
@@ -44,26 +49,20 @@
 
     public static final String NAME = "wrapper";
     private final byte[] source;
-    private final int offset;
-    private final int length;
-    static final WrapperQueryBuilder PROTOTYPE = new WrapperQueryBuilder(null, -1, -1);
+    static final WrapperQueryBuilder PROTOTYPE = new WrapperQueryBuilder(new byte[0]);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: we can just pass in null here?

File: docs/reference/migration/migrate_2_0.asciidoc [hunk 9]
```diff
@@ -138,6 +138,8 @@ In addition some query builders have been removed or renamed:
 * `filtered(...)` removed. Use `filteredQuery(...)` instead.
 * `inQuery(...)` removed.
 
+The constructor WrapperQueryBuilder(byte[] source, int offset, int length) has been removed, use WrapperQueryBuilder(byte[] source) instead.
+
```
Comment (javanna) [inline docs/reference/migration/migrate_2_0.asciidoc]: we need to add here the signature change to `public static WrapperQueryBuilder wrapperQuery(byte[] source)` in QueryBuiders.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java [hunk 10]
```diff
@@ -41,28 +39,33 @@ public WrapperQueryParser() {
...
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if ("query".equals(currentFieldName)) {
+                    source = parser.binaryValue();
+                } else {
+                    throw new QueryParsingException(parseContext, "[wrapper] query does not support [" + currentFieldName + "]");
+                }
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java]: I find it hard to read the diff here, I think the parser should go back to where it was before you introduced parsing of queryName and boost, which was removed afterwards.

File: core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        assertThat(wrapperQueryBuilder.validate().validationErrors().size(), is(1));
+    }
+
+    @Test
+    public void testEmptyWrappedQuery() throws IOException {
+        WrapperQueryBuilder wrapperQueryBuilder = new WrapperQueryBuilder("");
+        QueryParseContext context = createContext();
+        context.setAllowUnmappedFields(true);
+        try {
+            wrapperQueryBuilder.toQuery(context);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java]: shouldn't we be testing this in validate?

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java (hunk 10 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java]: I know but it is still a cleaner way to parse. So should I still revert the change?

File: core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java (hunk 11 above)
Comment (alexksikes) [inline core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java]: Probably as this is an error, but the parser did not originally check for it. You're call.

**Comment by alexksikes** [comment]
@javanna I addressed the comments and rebased. You can take a look. Thank you.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java (hunk 10 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java]: queryName and boost are not supported by this query, it makes no sense to parse them, hence I think we should go back to how we did things before here. The diff should get much smaller then.

File: core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java (hunk 11 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java]: this should happen in validate, and toQuery should simply rely on validate checks

File: docs/reference/migration/migrate_2_0.asciidoc [hunk 12]
```diff
@@ -138,6 +138,9 @@ In addition some query builders have been removed or renamed:
 * `filtered(...)` removed. Use `filteredQuery(...)` instead.
 * `inQuery(...)` removed.
 
+Also QueryBuilders.WrapperQueryBuilder(byte[] source, int offset, int length) has been removed,
```
Comment (javanna) [inline docs/reference/migration/migrate_2_0.asciidoc]: the method names start with a lowercase `w`. Shouldn't we have both changes here? QueryBuilders and WrapperQueryBuilder constructor? those are both public.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java [hunk 13]
```diff
@@ -71,19 +70,74 @@ public WrapperQueryBuilder(byte[] source, int offset, int length) {
...
 
     @Override
-    public String getName() {
-        return NAME;
+    protected Query doToQuery(QueryParseContext parseContext) throws IOException {
+        try (XContentParser qSourceParser = XContentFactory.xContent(source).createParser(source)) {
+            final QueryParseContext context = new QueryParseContext(parseContext.index(), parseContext.indexQueryParserService());
+            context.reset(qSourceParser);
+            Query query = context.parseInnerQuery();
+            parseContext.combineNamedQueries(context);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: parsing in toQuery feels very weird here... it will be a problem especially when we split the contexts into "parsing" and "to_query" because we will need the former here while we will have only the latter. @cbuescher what do you think we should do here instead? I am not sure yet....

**Comment by javanna** [comment]
I did another round, left some comments, also we need some discussion on when we effectively parse the binary query here...

File: docs/reference/migration/migrate_2_0.asciidoc (hunk 12 above)
Comment (alexksikes) [inline docs/reference/migration/migrate_2_0.asciidoc]: sure I'll make that change

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java (hunk 13 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: The doToQuery is probably going to change. In the mean time I don't see any other way of doing this.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java (hunk 10 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java]: Yes but the parsing goes on with the parsing of the json query. So I think we do have to modify the parsing, because the json query is parsed only later in toQuery.

File: core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java (hunk 11 above)
Comment (alexksikes) [inline core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java]: sure i'll add the check in validate then.

**Comment by alexksikes** [comment]
@javanna I rebased it to incorporate the latest changes. Something I had to do is move supportsBoostAndQueryName() method to AbstractQueryBuilder. Please see comment in code. Thank you.

File: core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java [hunk 14]
```diff
@@ -79,6 +79,16 @@ public final Query toQuery(QueryParseContext parseContext) throws IOException {
         return query;
     }
 
+    /**
+     * Few queries allow you to set the boost and queryName but don't do anything with it. This method allows
+     * to disable boost and queryName for those queries.
+     */
+    //norelease this has to be put in here because of the call on toQuery which would override the boost and queryName
+    //norelease of the wrapped query with defaults
+    protected boolean supportsBoostAndQueryName() {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: //norelease is used for code that can be removed later on in the refactoring. I don't think it should be used here, or at least I don't see how we can remove this method if we introduce it here. The idea for `supportsBoostAndQueryName` was to be only a test method to "mark" few exceptional queries, it wasn't meant to be moved to prod code ever.

The reason why you did this is because the wrapped query has its own boost that gets overwritten by the default 1.0f boost of the outer query. I tend to think that this is more a test problem, and the doAssertLuceneQuery in WrapperQueryBuilderTest should be adjusted. Maybe there are other ways or ideas, but we should not move this `supportsBoostAndQueryName` to prod code I think.

**Comment by javanna** [comment]
I had a look, left a comment, thinking more about it this query will give us headache, especially in relation to #12527. This is the only query that needs a `QueryParseContext` in `toQuery`, along with the `QueryShardContext`. We will need to discuss this with @cbuescher and see what we can do about it.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java (hunk 13 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: ok there is not much that we can do here, we have to parse this query in the doToQuery. When you rebase you will get a QueryShardContext here as an argument. I would create a new QueryParseContext from it so you can parse the inner query.

**Comment by javanna** [comment]
@alexksikes could you please rebase and address comments?

File: core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java (hunk 14 above)
Comment (alexksikes) [inline core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: AbstractQueryBuilder#toQuery does set a boost unless it is disabled. If I remember correctly I think that was the idea of being to disable boost as part of the actual query builder, not just in the test.

**Comment by alexksikes** [comment]
@javanna I rebased and addressed the comments. Thanks for the review.

File: core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java (hunk 14 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: I am on the fence on this. We have one query with `boost: 2.0` which holds an inner query with `boost: 3.0`. Let's say that the final lucene query will be the lucene representation of the inner query. What about the final boost, should it be 2.0 or 3.0? Right now the boost set on the main query overrides the inner boost. I am not sure if this is the right behaviour, which would simply require to adjust the doAssertLuceneQuery here to take this behaviour into account. But `supportsBoostAnQueryName` should not be moved to prod code. I would go back to having it in test and I would adjust the doAssertLuceneQuery here to work properly without needing to disable anything. Then we can open an issue and discuss what the final boost should be.

File: core/src/test/java/org/elasticsearch/index/query/QueryFilterBuilderTest.java [hunk 15]
```diff
@@ -52,11 +52,6 @@ protected void doAssertLuceneQuery(QueryFilterBuilder queryBuilder, Query query,
         }
     }
 
-    @Override
-    protected boolean supportsBoostAndQueryName() {
-        return false;
-    }
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/QueryFilterBuilderTest.java]: why was this removed?

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java [hunk 16]
```diff
@@ -86,4 +92,44 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
...
         return NAME;
     }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        try (XContentParser qSourceParser = XContentFactory.xContent(source).createParser(source)) {
+            final QueryParseContext parseContext = new QueryParseContext(context);
+            parseContext.reset(qSourceParser);
+            Query query = parseContext.parseInnerQuery();
+            context.combineNamedQueries(context);  // NO COMMIT: not really sure if this is the right way
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: I had a look at this, I think we can do the following:

```
QueryBuilder queryBuilder = parseContext.parseInnerQueryBuilder();
 return queryBuilder.toQuery(context);
```

Given that we reuse the original query shard context, I think we my not need to call combineNamedQueries, actually we could remove that method cause this is the only reason why we have it in the first place. What do you think @cbuescher ?

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java]: Yes, I think you are right. Previously in the parser, the context to parse the wrapped query was a copy with empty namedQueries, to whatever got added there during the inner parsing needed to be re-added to the actual context. We now add all the named queries in AbstractQueryBuilder#toQuery(), so I think when we use the original shard context here, all named queries generated by the inner query builder are added there too.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java (hunk 10 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java]: yea  the `parser.binaryValue()` that you introduced instead of parsing the inner query is fine. But I don't think we need a loop if we parse a single field, that's what I meant when I say to back to what we did before.

File: core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java (hunk 14 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: I see that you moved back to supportsBoostAndQueryName in BaseQUeryTestCase. The way you do it now makes perfect sense.

**Comment by javanna** [comment]
did another round, left a few comments

**Comment by alexksikes** [comment]
@javanna comments addressed. Thanks for the review.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java (hunk 10 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java]: I don't see this comment addressed. Once again, we don't need a loop if we read a single field. The way we parsed before was simpler, can we go back please?

File: core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @Override
+    protected void doAssertLuceneQuery(WrapperQueryBuilder queryBuilder, Query query, QueryShardContext context) throws IOException {
+        try (XContentParser qSourceParser = XContentFactory.xContent(queryBuilder.source()).createParser(queryBuilder.source())) {
+            final QueryParseContext parseContext = new QueryParseContext(context);
+            parseContext.reset(qSourceParser);
+            Query expected = parseContext.parseInnerQueryBuilder().toQuery(context);
+            if (expected != null) {
+                expected.setBoost(queryBuilder.boost());
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/WrapperQueryBuilderTest.java]: can you replace with `expected.setBoost(AbstractQueryBuilder.DEFAULT_BOOST)` for clarity? I will also open up an issue for discussion on this boost matter.

**Comment by javanna** [comment]
I did a final round, one old comment on parsing wasn't addressed yet, plus left a minor comment on testing.

**Comment by alexksikes** [comment]
@javanna addressed the comments and rebased. Thanks for the review.

File: core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java [hunk 18]
```diff
@@ -55,15 +52,14 @@ public Query parse(QueryShardContext context) throws IOException, QueryParsingEx
...
-            Query result = contextCopy.parseContext().parseInnerQuery();
-            parser.nextToken();
-            context.combineNamedQueries(contextCopy);
-            return result;
+        byte[] source = parser.binaryValue();
+
+        parser.nextToken();
+
+        if (source == null) {
+            throw new QueryParsingException(parseContext, "wrapper query has no [source] specified");
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java]: I think the field name should be `query` in the error message

**Comment by javanna** [comment]
LGTM besides the tiny comment I left

**Comment by javanna** [comment]
This change is breaking for the java API as it removes the `WrapperQueryBuilder(byte[] source, int offset, int length)` constructor, use `WrapperQueryBuilder(byte[] source)` instead.
